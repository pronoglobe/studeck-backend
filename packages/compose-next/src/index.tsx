import * as React from 'react';

export const addInitialProps = (getInitialProps: (ctx: any) => any) => {
  // tslint:disable-next-line:variable-name
  return (Component: React.ComponentType) =>
    class extends React.Component {
      public static getInitialProps(ctx: any): any {
        return getInitialProps(ctx);
      }
      public render(): JSX.Element {
        return <Component {...this.props} />;
      }
    };
};

export type ComponentWithInitialProps = React.ComponentType<any> & {
  getInitialProps?: (ctx?: any) => any;
};

// tslint:disable-next-line:variable-name
export const composeInitialProps = (Parent: ComponentWithInitialProps) => (Child: ComponentWithInitialProps) =>
  class extends React.Component<any> {
    public static async getInitialProps(ctx?: any): Promise<{ parentProps: any; childProps: any }> {
      return {
        parentProps: Parent.getInitialProps ? await Parent.getInitialProps(ctx) : {},
        childProps: Child.getInitialProps ? await Child.getInitialProps(ctx) : {},
      };
    }
    public render(): JSX.Element {
      return (
        <Parent {...this.props} {...this.props.parentProps}>
          <Child {...this.props} {...this.props.childProps} />
        </Parent>
      );
    }
  };
