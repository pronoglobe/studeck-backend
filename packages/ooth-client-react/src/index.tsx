import * as React from 'react';
import { getContext } from 'recompose';
import { OothClient, User } from 'ooth-client';
import * as PropTypes from 'prop-types';

type Props = {
  client: OothClient;
  initialUser: User | undefined;
};

type Status = { user: User | undefined };

export class OothProvider extends React.Component<Props, Status> {
  public static childContextTypes: any = {
    oothClient: PropTypes.object.isRequired,
    user: PropTypes.object,
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      user: props.initialUser,
    };
  }

  public componentDidMount(): void {
    this.props.client.start().then((user) => {
      this.setState({
        user,
      });
      this.props.client.on('user', this.onUser);
    });
  }

  public componentWillUnmount(): void {
    this.props.client.unsubscribe('user', this.onUser);
  }

  public render(): JSX.Element | null {
    if (!this.props.children) {
      return null;
    }
    return React.Children.only(this.props.children);
  }

  public getChildContext(): any {
    const { client: oothClient } = this.props;
    const { user } = this.state;
    return {
      oothClient,
      user,
    };
  }

  private onUser = async (user: User) => {
    this.setState({
      user,
    });
  };
}

export const withOoth = getContext({
  oothClient: PropTypes.object.isRequired,
});

export const withUser = getContext({
  user: PropTypes.object,
});
