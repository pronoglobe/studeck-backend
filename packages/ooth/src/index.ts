import * as bodyParser from 'body-parser';
import { Application, Request, RequestHandler, Response, Router } from 'express';
import { getI18n, Translations, Values } from 'ooth-i18n';
import * as passport from 'passport';
import * as expressSession from 'express-session';
import * as cookieParser from 'cookie-parser';

const { Strategy } = require('passport-custom');
const locale = require('locale');

export interface Backend {
  getUserById(id: string): Promise<User>;
  getUser(fields: { [key: string]: any }): Promise<User>;
  getUserByValue(fields: string[], value: any): Promise<User>;
  updateUser(id: string, fields: { [key: string]: any }): Promise<any>;
  insertUser(fields: { [key: string]: any }): Promise<any>;
}

export type Options = {
  app: Application;
  backend: Backend;
  path?: string;
  sessionSecret?: string;
  session?: RequestHandler;

  defaultLanguage?: string;
  translations?: Translations;
};

export type StrategyValues = {
  [field: string]: any;
};

export type User = {
  _id: string;
  [strategyName: string]: undefined | string | Date | StrategyValues;
};

type Condition = (req: FullRequest) => boolean;

type Method<T> = (params: any, user: string | undefined, locale: string) => Promise<T>;

type Strategy = {
  profileFields: {
    [key: string]: true;
  };
  methods: {
    [key: string]: Method<any>;
  };
  uniqueFields: {
    [key: string]: true;
  };
  secondaryAuth: {
    [key: string]: Condition;
  };
};

type Strategies = {
  [key: string]: Strategy;
};

type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

type Override<T, K> = Omit<T, keyof K> & K;

export type FullRequest = Override<
  Request,
  {
    locale: string;
    user?: string | undefined;
    session?: {
      id: string;
      cookie: {
        maxAge?: number;
      };
    };
  }
>;

type MyRequestHandler = (req: FullRequest, res: Response, next: () => any) => any;

type UserIdResolver<T> = (req: FullRequest, authResult: T) => Promise<string | undefined>;

type Result = { [key: string]: any };

type Afterware = ((res: Result, userId: string | undefined, req: FullRequest, response: Response) => Promise<Result>);

const DEFAULT_LANGUAGE = 'en';
const DEFAULT_TRANSLATIONS = {
  en: require('../i18n/en.json'),
};

const authenticate = <T>(methodName: string, session: boolean, req: FullRequest, res: Response): Promise<T> =>
  new Promise((resolve, reject) =>
    passport.authenticate(
      methodName,
      {
        session,
      },
      (err: Error, user: any, info: any) =>
        err ? reject(err) : !user && info && info.message ? reject(info.message) : resolve(user),
    )(req, res),
  );

type EventListener = (payload: any) => Promise<void>;

export class Ooth {
  private path: string;
  private uniqueFields: { [key: string]: { [key: string]: string } } = {};
  private strategies: Strategies = {};
  private route: Router = Router();
  private translations: Translations;
  private defaultLanguage: string;
  private __: (s: string, values: Values | null, language: string) => string;
  private backend: Backend;
  private app: Application;
  private middleware: RequestHandler[] = [];
  private afterware: Afterware[] = [];
  private authAfterware: Afterware[] = [];
  private session?: RequestHandler;
  private listeners: { [name: string]: EventListener[] } = {};

  constructor({ app, backend, path, defaultLanguage, translations, session, sessionSecret }: Options) {
    if (!app) {
      throw new Error('App is required.');
    }
    this.app = app;

    if (!backend) {
      throw new Error('Backend is required.');
    }
    this.backend = backend;
    this.path = path || '/';
    this.translations = translations || DEFAULT_TRANSLATIONS;
    this.defaultLanguage = defaultLanguage || DEFAULT_LANGUAGE;
    this.__ = getI18n(this.translations, this.defaultLanguage);
    if (sessionSecret && session) {
      throw new Error('Either set session or sessionSecret, not both');
    }
    if (session) {
      this.session = session;
    }
    if (sessionSecret) {
      this.session = expressSession({
        secret: sessionSecret,
        name: 'ooth-session-id',
        resave: false,
        saveUninitialized: true,
      });
    }
    if (this.session) {
      this.registerAuthAfterware(async (result: Result, _userId: string | undefined, req: FullRequest) => {
        if (req.body.remember) {
          req.session!.cookie.maxAge = 30 * 24 * 60 * 60 * 1000;
        }
        return result;
      });
    }

    // App-wide configuration
    this.app.use(bodyParser.json());
    this.app.use(locale(Object.keys(this.translations), this.defaultLanguage));
    if (this.session) {
      this.app.use(cookieParser());
      this.app.use(this.session);
    }
    this.app.use(passport.initialize());
    if (this.session) {
      this.app.use(passport.session());
      passport.serializeUser((id: string | undefined, done: (e: Error | null, id: string | undefined) => void) =>
        done(null, id),
      );
      passport.deserializeUser((id: string | undefined, done: (e: Error | null, id: string | undefined) => void) =>
        done(null, id),
      );
      this.registerPrimaryAuth(
        'session',
        'logout',
        [this.requireLogged],
        new Strategy((_req: FullRequest, done: (e: Error | null, v: null | User | StrategyValues) => void) =>
          done(null, null),
        ),
      );
    }
    this.app.use(async (req, res, next) => {
      try {
        const fullRequest: FullRequest = req as any;
        for (const strategyName of Object.keys(this.strategies)) {
          for (const methodName of Object.keys(this.strategies[strategyName].secondaryAuth)) {
            if (this.strategies[strategyName].secondaryAuth[methodName](fullRequest)) {
              const userId = await authenticate<string | undefined>(
                `${strategyName}-${methodName}`,
                !!this.session,
                fullRequest,
                res,
              );
              await new Promise((res, rej) =>
                fullRequest.login(userId, { session: !!this.session }, (e) => (e ? rej(e) : res())),
              );
            }
          }
        }
        next();
      } catch (e) {
        console.error(e);
        return res.status(400).send({
          status: 'error',
          message: typeof e === 'string' ? e : e.message,
        });
      }
    });
    this.app.use(this.path, this.route);
  }

  public usesSession(): boolean {
    return !!this.session;
  }

  public registerUniqueField(strategyName: string, id: string, fieldName: string): void {
    if (!this.uniqueFields[id]) {
      this.uniqueFields[id] = {};
    }
    this.uniqueFields[id][strategyName] = fieldName;
  }

  public registerStrategyUniqueField(strategyName: string, ...fieldNames: string[]): void {
    for (const fieldName of fieldNames) {
      this.getStrategy(strategyName).uniqueFields[fieldName] = true;
    }
  }

  public registerProfileFields(strategyName: string, ...fieldNames: string[]): void {
    for (const fieldName of fieldNames) {
      this.getStrategy(strategyName).profileFields[fieldName] = true;
    }
  }

  public getUniqueField = (user: User, fieldName: string) => {
    if (this.uniqueFields[fieldName]) {
      for (const strategyName of Object.keys(this.uniqueFields[fieldName])) {
        const actualFieldName = this.uniqueFields[fieldName][strategyName];
        if (user[strategyName] && (user[strategyName] as StrategyValues)[actualFieldName]) {
          return (user[strategyName] as StrategyValues)[actualFieldName];
        }
      }
    }
    return null;
  };

  public getUserByFields = async (strategyName: string, fields: { [key: string]: any }) => {
    const actualFields: { [key: string]: any } = {};
    Object.keys(fields).forEach((field) => {
      actualFields[`${strategyName}.${field}`] = fields[field];
    });
    return await this.backend!.getUser(actualFields);
  };

  public requireLogged = (req: FullRequest, res: Response, next: () => any) => {
    if (!req.user) {
      return res.status(400).send({
        status: 'error',
        message: this.__('not_logged', null, req.locale),
      });
    }
    next();
  };

  public requireNotLogged = (req: FullRequest, res: Response, next: () => any) => {
    if (req.user) {
      return res.status(400).send({
        status: 'error',
        message: this.__('already_logged', null, req.locale),
      });
    }
    next();
  };

  public requireNotRegistered = (req: FullRequest, res: Response, next: () => any) => {
    if (req.user) {
      return res.status(400).send({
        status: 'error',
        message: this.__('already_registered', null, req.locale),
      });
    }
    next();
  };

  public requireRegisteredWith = (strategy: string) => {
    return (req: FullRequest, res: Response, next: () => any) =>
      this.requireLogged(req, res, async () => {
        const user = await this.getUserById(req.user!);
        if (!user[strategy]) {
          return res.status(400).send({
            status: 'error',
            message: this.__('not_registered_with', { strategy }, req.locale),
          });
        }
        next();
      });
  };

  public registerPrimaryAuth<T>(
    strategyName: string,
    method: string,
    middleware: MyRequestHandler[],
    strategy: passport.Strategy,
    resolveUserId: UserIdResolver<T> = async (_req: FullRequest, userId: T) => userId as any,
  ): void {
    const methodName = `${strategyName}-${method}`;
    const routeName = `/${strategyName}/${method}`;
    passport.use(methodName, strategy);
    this.route.post(routeName, ...((middleware as any) as RequestHandler[]), async (req: Request, res: Response) => {
      const fullRequest: FullRequest = req as any;
      try {
        const authResult = await authenticate<T>(methodName, !!this.session, fullRequest, res);
        const userId = await resolveUserId(fullRequest, authResult);
        if (req.user !== userId) {
          if (userId) {
            this.emit('ooth', 'login', { userId, sessionId: req.session && req.session.id });
            await new Promise((res, rej) =>
              fullRequest.login(userId, { session: !!this.session }, (e) => (e ? rej(e) : res())),
            );
          } else {
            this.emit('ooth', 'logout', { userId: req.user, sessionId: req.session && req.session.id });
            req.logout();
          }
        }

        let result = {};
        for (const aw of [...this.authAfterware, ...this.afterware]) {
          result = await aw(result, fullRequest.user, fullRequest, res);
        }
        return res.send(result);
      } catch (e) {
        console.error(e);
        return res.status(400).send({
          status: 'error',
          message: e.message || e,
        });
      }
    });
  }

  public registerPrimaryConnect(
    strategyName: string,
    method: string,
    middleware: MyRequestHandler[],
    strategy: passport.Strategy,
  ): void {
    this.registerPrimaryAuth<StrategyValues | null>(
      strategyName,
      method,
      middleware,
      strategy,
      async (req: FullRequest, userPart: StrategyValues | null) => {
        if (!userPart) {
          return;
        }

        let userId: string | undefined;
        for (const field of Object.keys(this.getStrategy(strategyName).uniqueFields)) {
          const value = userPart[field];
          if (value) {
            const userCandidate = await this.backend!.getUser({
              [`${strategyName}.${field}`]: value,
            });
            if (userCandidate) {
              if (!userId || userId === userCandidate._id) {
                userId = userCandidate._id;
              } else {
                throw new Error(this.__('register.ambiguous_authentication_strategy', null, req.locale));
              }
            }
          }
        }
        for (const field of Object.keys(this.uniqueFields)) {
          if (this.uniqueFields[field][strategyName]) {
            const value = userPart[this.uniqueFields[field][strategyName]];
            if (value) {
              const userCandidate = await this.getUserByUniqueField(field, value);
              if (userCandidate) {
                if (!userId || userId === userCandidate._id) {
                  userId = userCandidate._id;
                } else {
                  throw new Error(this.__('register.ambiguous_authentication_strategy', null, req.locale));
                }
              }
            }
          }
        }

        let registered = false;
        if (req.user) {
          // User is already logged in

          if (userId && userId !== req.user) {
            throw new Error(this.__('register.foreign_authentication_strategy', null, req.locale));
          }

          userId = req.user;

          // Update user
          await this.updateUser(strategyName, userId, userPart);
        } else {
          // User hasn't logged in

          if (userId) {
            // User exists already, update
            await this.updateUser(strategyName, userId, userPart);
          } else {
            // Need to create a new user
            userId = await this.insertUser(strategyName, userPart);
            registered = true;
          }
        }

        if (registered) {
          this.emit('ooth', 'register', { userId });
        }

        return userId;
      },
    );
  }

  public registerSecondaryAuth(
    strategyName: string,
    method: string,
    condition: Condition,
    strategy: passport.Strategy,
  ): void {
    const methodName = `${strategyName}-${method}`;
    if (this.getStrategy(strategyName).secondaryAuth[method]) {
      throw new Error(`Secondary auth ${methodName} has already been registered.`);
    }
    passport.use(methodName, strategy);
    this.getStrategy(strategyName).secondaryAuth[method] = condition;
  }

  public registerMethod<T>(
    strategyName: string,
    method: string,
    middleware: MyRequestHandler[],
    handler: Method<T>,
    httpMethod: 'POST' | 'GET' = 'POST',
  ): void {
    this.getStrategy(strategyName).methods[method] = handler;

    const finalHandler = async (req: FullRequest, res: Response) => {
      try {
        let result = await this.callMethod(strategyName, method, req.body, req.user, req.locale);
        for (const aw of this.afterware) {
          result = await aw(result, req.user, req, res);
        }
        res.send(result);
      } catch (e) {
        console.error(e);
        return res.status(400).send({
          status: 'error',
          message: e.message,
        });
      }
    };

    this.route[httpMethod === 'GET' ? 'get' : 'post'](
      `/${strategyName}/${method}`,
      ...((middleware as any) as RequestHandler[]),
      (finalHandler as any) as RequestHandler,
    );
  }

  public async callMethod<T>(strategyName: string, method: string, params: any, user: any, locale: string): Promise<T> {
    return await this.getStrategy(strategyName).methods[method](params, user, locale);
  }

  public registerMiddleware(...middleware: RequestHandler[]): void {
    this.middleware.push(...middleware);
  }

  public registerAfterware(...afterware: Afterware[]): void {
    this.afterware.push(...afterware);
  }

  public registerAuthAfterware(...afterware: Afterware[]): void {
    this.authAfterware.push(...afterware);
  }

  public getUserById = async (id: string) => {
    return this.backend.getUserById(id);
  };

  public getUserByUniqueField = async (fieldName: string, value: any) => {
    return await this.backend!.getUserByValue(
      Object.keys(this.uniqueFields[fieldName]).map(
        (strategyName) => `${strategyName}.${this.uniqueFields[fieldName][strategyName]}`,
      ),
      value,
    );
  };

  public updateUser = async (strategyName: string, id: string, fields: StrategyValues) => {
    const actualFields: { [key: string]: string } = {};
    Object.keys(fields).forEach((field) => {
      actualFields[`${strategyName}.${field}`] = fields[field];
    });
    return await this.backend!.updateUser(id, actualFields);
  };

  public getProfile(user: User | undefined): User | undefined {
    if (!user) {
      return;
    }
    const profile: User = {
      _id: user._id,
    };
    for (const strategyName of Object.keys(this.strategies)) {
      if (!user[strategyName]) {
        continue;
      }

      for (const fieldName of Object.keys(this.getStrategy(strategyName).profileFields)) {
        if (!profile[strategyName]) {
          profile[strategyName] = {};
        }
        (profile[strategyName] as StrategyValues)[fieldName] = (user[strategyName] as StrategyValues)[fieldName];
      }
    }
    return profile;
  }

  public insertUser = async (strategyName: string, fields: StrategyValues) => {
    const query: { [key: string]: { [key: string]: any } } = {};
    if (fields) {
      query[strategyName] = fields;
    }
    return await this.backend!.insertUser(query);
  };

  public getApp(): Application {
    return this.app;
  }

  public getRoute(): Router {
    return this.route;
  }

  public on(strategyName: string, eventName: string, eventListener: EventListener): void {
    const fullEventName = `${strategyName}--${eventName}`;
    if (!this.listeners[fullEventName]) {
      this.listeners[fullEventName] = [];
    }

    this.listeners[fullEventName].push(eventListener);
  }

  public async emit(strategyName: string, eventName: string, payload: any): Promise<void> {
    const fullEventName = `${strategyName}--${eventName}`;
    if (this.listeners[fullEventName]) {
      for (const listener of this.listeners[fullEventName]) {
        await listener(payload);
      }
    }
  }

  private getStrategy(strategyName: string): Strategy {
    if (!this.strategies[strategyName]) {
      this.strategies[strategyName] = {
        methods: {},
        profileFields: {},
        uniqueFields: {},
        secondaryAuth: {},
      };
    }

    return this.strategies[strategyName];
  }
}
