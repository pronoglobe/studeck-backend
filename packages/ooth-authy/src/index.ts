import { FullRequest, Ooth, StrategyValues } from 'ooth';
import { callbackify } from 'util';

const fetch = require('node-fetch');

const { Strategy } = require('passport-custom');

type Config = {
  name?: string;
  ooth: Ooth;
  apiKey: string;
};

export default function({ name = 'authy', ooth, apiKey }: Config): void {
  ooth.registerUniqueField(name, 'email', 'email');
  ooth.registerProfileFields(name, 'email');
  ooth.registerStrategyUniqueField(name, 'id');

  ooth.registerMethod(name, 'register', [ooth.requireNotLogged], async ({ email, cellphone, country_code }) => {
    if (!email || typeof email !== 'string') {
      throw new Error('Email is required.');
    }
    if (!cellphone || typeof cellphone !== 'string') {
      throw new Error('Invalid phone number.');
    }
    if (!country_code || typeof country_code !== 'string') {
      throw new Error('Invalid coutnry code.');
    }

    const existingUser = await ooth.getUserByUniqueField('email', email);

    if (existingUser) {
      throw new Error('This email is already registered');
    }

    const res = await fetch(`https://api.authy.com/protected/json/users/new`, {
      method: 'POST',
      body: `user[email]=${email}&user[cellphone]=${cellphone}&user[country_code]=${country_code}`,
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'X-Authy-API-Key': apiKey,
      },
    });
    const data = await res.json();
    if (data.errors) {
      console.error(data);
      const key = Object.keys(data.errors)[0];
      throw new Error(`${key} ${data.errors[key]}`);
    }

    await ooth.insertUser(name, {
      email,
      cellphone,
      country_code,
      id: data.user.id,
    });

    return {
      message: 'Registered successfully',
    };
  });

  ooth.registerMethod(name, 'otp', [ooth.requireNotLogged], async ({ email }) => {
    const user = await ooth.getUserByUniqueField('email', email);
    if (!user || !user[name]) {
      throw new Error('This email is not registered with this authentication strategy.');
    }
    const values = user[name] as StrategyValues;
    const res = await fetch(`https://api.authy.com/protected/json/sms/${values.id}`, {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'X-Authy-API-Key': apiKey,
      },
    });
    const data = await res.json();
    if (data.errors) {
      throw new Error(data.errors[Object.keys(data.errors)[0]]);
    }

    return {
      message: 'A one-time password has been send to you via SMS.',
    };
  });

  ooth.registerPrimaryAuth(
    name,
    'login',
    [ooth.requireNotLogged],
    new Strategy(
      callbackify(async (req: FullRequest) => {
        const { email, token } = req.body;
        if (!email || typeof email !== 'string') {
          throw new Error('Invalid email.');
        }

        if (!token || (typeof token !== 'string' && typeof token !== 'number')) {
          throw new Error('Invalid token.');
        }

        const user = await ooth.getUserByUniqueField('email', email);

        if (!user || !user[name]) {
          throw new Error('This email is not registered with this authentication strategy.');
        }
        const values = user[name] as StrategyValues;

        console.log(`https://api.authy.com/protected/json/verify/${token}/${values.id}`);

        const res = await fetch(`https://api.authy.com/protected/json/verify/${token}/${values.id}`, {
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
            'X-Authy-API-Key': apiKey,
          },
        });
        const data = await res.json();
        if (data.errors) {
          throw new Error(data.errors[Object.keys(data.errors)[0]]);
        }

        return user._id;
      }),
    ),
  );
}
