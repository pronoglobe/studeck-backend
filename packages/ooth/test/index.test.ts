import * as express from 'express';
import { cloneDeep } from 'lodash';
import { Db, MongoClient } from 'mongodb';
import MongodbMemoryServer from 'mongodb-memory-server';
import { OothMongo } from 'ooth-mongo';
import * as request from 'request-promise';

import { Ooth, User } from '../src';
import { Server } from 'http';

const { Strategy } = require('passport-custom');

let mongoServer: MongodbMemoryServer;
let con: MongoClient;
let db: Db;
let app: express.Express;
let server: Server;
let ooth: Ooth;
let oothMongo: OothMongo;

// May require additional time for downloading MongoDB binaries
jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

const startServer = () => {
  return new Promise((resolve) => {
    server = app.listen(8080, resolve);
  });
};

const obfuscate = (obj: any, ...paths: string[]) => {
  const res = cloneDeep(obj);
  for (const path of paths) {
    const keys = path.split('.');
    let current = res;
    for (const key of keys.slice(0, -1)) {
      current = current[key];
    }
    current[keys[keys.length - 1]] = '<obfuscated>';
  }

  return res;
};

describe('ooth', () => {
  beforeAll(async () => {
    mongoServer = new MongodbMemoryServer();
    const connectionString = await mongoServer.getConnectionString();
    const dbName = await mongoServer.getDbName();
    con = await MongoClient.connect(connectionString);
    db = con.db(dbName);
  });

  afterAll(async () => {
    await con.close();
    await mongoServer.stop();
  });

  describe('no session', () => {
    beforeEach(() => {
      app = express();
      oothMongo = new OothMongo(db);
      ooth = new Ooth({
        app,
        backend: oothMongo,
        path: '',
        translations: {
          en: require('../i18n/en.json'),
          fr: {
            not_logged: 'French not logged',
          },
        },
      });
      ooth.registerAfterware(async (result: { [key: string]: any }, userId: string | undefined) => {
        if (userId) {
          result.user = ooth.getProfile(await ooth.getUserById(userId));
        }

        return result;
      });
    });

    afterEach(async () => {
      if (server) {
        await server.close();
      }
      if (db) {
        await db.dropDatabase();
      }
    });

    describe('methods', () => {
      test('handle method', async () => {
        ooth.registerMethod<{ message: string }>('test', 'foo', [], async () => ({
          message: 'hi',
        }));

        await startServer();
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/test/foo',
          json: true,
        });
        expect(res).toMatchSnapshot();
      });

      test('can call method', async () => {
        ooth.registerMethod<{ message: string }>('test', 'foo', [], async (body) => body);

        const res = await ooth.callMethod('test', 'foo', 'hello world', null, 'en');
        expect(res).toBe('hello world');
      });

      test('fails with requireLogged', async () => {
        ooth.registerMethod('test', 'foo', [ooth.requireLogged], async () => ({
          message: 'hi',
        }));

        await startServer();
        try {
          await request({
            method: 'POST',
            uri: 'http://localhost:8080/test/foo',
            json: true,
          });
        } catch (e) {
          expect(e.response.body).toMatchSnapshot();
          return;
        }
        throw new Error("Didn't fail");
      });

      test('translates error', async () => {
        ooth.registerMethod('test', 'foo', [ooth.requireLogged], async () => ({
          message: 'hi',
        }));

        await startServer();
        try {
          await request({
            method: 'POST',
            uri: 'http://localhost:8080/test/foo',
            headers: {
              'Accept-Language': 'fr',
            },
            json: true,
          });
        } catch (e) {
          expect(e.response.body).toMatchSnapshot();
          return;
        }
        throw new Error("Didn't fail");
      });

      test('handle errors', async () => {
        ooth.registerMethod('test', 'foo', [], () => {
          throw new Error('Error message.');
        });

        await startServer();
        try {
          await request({
            method: 'POST',
            uri: 'http://localhost:8080/test/foo',
            json: true,
          });
          expect(false);
        } catch (e) {
          expect(e.response.body).toMatchSnapshot();
        }
      });

      test('handle async errors', async () => {
        ooth.registerMethod('test', 'foo', [], async () =>
          Promise.resolve().then(() => {
            throw new Error('Error message.');
          }),
        );

        await startServer();
        try {
          await request({
            method: 'POST',
            uri: 'http://localhost:8080/test/foo',
            json: true,
          });
          expect(false);
        } catch (e) {
          expect(e.response.body).toMatchSnapshot();
        }
      });
    });

    describe('connect method', () => {
      beforeEach(async () => {
        ooth.registerProfileFields('test', 'foo');
        ooth.registerUniqueField('test', 'bar', 'bar');
        ooth.registerPrimaryConnect('test', 'login', [], new Strategy((req: any, done: any) => done(null, req.body)));
        ooth.registerProfileFields('test2', 'baz');
        ooth.registerUniqueField('test2', 'bar', 'bar');
        ooth.registerPrimaryConnect('test2', 'login', [], new Strategy((req: any, done: any) => done(null, req.body)));
        await startServer();
      });

      test('can register', async () => {
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/test/login',
          body: {
            foo: 1,
            bar: 2,
          },
          json: true,
        });
        expect(obfuscate(res.user, '_id')).toMatchSnapshot();
      });

      test('can log in', async () => {
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/test/login',
          body: {
            foo: 1,
            bar: 2,
          },
          json: true,
        });
        expect(obfuscate(res, 'user._id')).toMatchSnapshot();
      });

      describe('after login', () => {
        let user: User;

        beforeEach(async () => {
          const res = await request({
            method: 'POST',
            uri: 'http://localhost:8080/test/login',
            body: {
              foo: 1,
              bar: 2,
            },
            json: true,
            resolveWithFullResponse: true,
          });
          user = res.body.user;
        });

        test('can log in again', async () => {
          const res = await request({
            method: 'POST',
            uri: 'http://localhost:8080/test/login',
            body: {
              foo: 2,
              bar: 2,
            },
            json: true,
          });
          expect(res.user._id).toBe(user._id);
          expect(obfuscate(res.user, '_id')).toMatchSnapshot();
        });

        test('can log in with other strategy', async () => {
          const res = await request({
            method: 'POST',
            uri: 'http://localhost:8080/test2/login',
            body: {
              foo: 2,
              bar: 2,
            },
            json: true,
          });
          expect(res.user._id).toBe(user._id);
          expect(obfuscate(res.user, '_id')).toMatchSnapshot();
        });

        test('can connect another strategy', async () => {
          ooth.registerSecondaryAuth('bla', 'bla', () => true, new Strategy((_req: any, done: any) => done(null, user._id)));
          const res = await request({
            method: 'POST',
            uri: 'http://localhost:8080/test2/login',
            body: {
              bar: 3,
              baz: 45,
            },
            json: true,
          });
          expect(res.user._id).toBe(user._id);
          expect(obfuscate(res.user, '_id')).toMatchSnapshot();
        });
      });
    });
  });

  describe('session', () => {
    beforeEach(async () => {
      app = express();
      oothMongo = new OothMongo(db);
      ooth = new Ooth({
        app,
        backend: oothMongo,
        path: '',
        sessionSecret: 'secret',
      });
      ooth.registerProfileFields('foo', 'x');
      ooth.registerPrimaryConnect('foo', 'login', [], new Strategy((_req: any, done: any) => done(null, { x: 1 })));
      ooth.registerMethod('foo', 'bar', [ooth.requireLogged], async () => ({ x: 1 }));
      ooth.registerAfterware(async (result: { [key: string]: any }, userId: string | undefined) => {
        if (userId) {
          result.user = ooth.getProfile(await ooth.getUserById(userId));
        }

        return result;
      });
      await startServer();
    });

    afterEach(async () => {
      if (server) {
        await server.close();
      }
      if (db) {
        await db.dropDatabase();
      }
    });

    describe('methods', () => {
      test('fails without cookies', async () => {
        try {
          await request({
            method: 'POST',
            uri: 'http://localhost:8080/foo/bar',
            json: true,
          });
        } catch (e) {
          expect(e.response.body).toMatchSnapshot();
          return;
        }
        throw new Error("Didn't throw");
      });

      test('can do secondary auth with cookies', async () => {
        const lres = await request({
          method: 'POST',
          uri: 'http://localhost:8080/foo/login',
          json: true,
          resolveWithFullResponse: true,
        });

        expect(obfuscate(lres.body.user, '_id')).toMatchSnapshot();

        expect(/Expires/.test(lres.headers['set-cookie'])).toBe(false);

        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/foo/bar',
          headers: {
            Cookie: lres.headers['set-cookie'],
          },
          json: true,
        });

        expect(obfuscate(res.user, '_id')).toMatchSnapshot();
      });

      test('remember', async () => {
        const lres = await request({
          method: 'POST',
          uri: 'http://localhost:8080/foo/login',
          json: true,
          body: { remember: true },
          resolveWithFullResponse: true,
        });

        expect(/Expires/.test(lres.headers['set-cookie'])).toBe(true);
      });

      test('can log out', async () => {
        const lres = await request({
          method: 'POST',
          uri: 'http://localhost:8080/foo/login',
          json: true,
          resolveWithFullResponse: true,
        });

        expect(obfuscate(lres.body.user, '_id')).toMatchSnapshot();

        const reslogout = await request({
          method: 'POST',
          uri: 'http://localhost:8080/session/logout',
          headers: {
            Cookie: lres.headers['set-cookie'],
          },
          json: true,
        });

        expect(reslogout).toMatchSnapshot();

        try {
          await request({
            method: 'POST',
            uri: 'http://localhost:8080/foo/bar',
            headers: {
              Cookie: lres.headers['set-cookie'],
            },
            json: true,
          });
        } catch (e) {
          expect(e.response.body).toMatchSnapshot();
          return;
        }
        throw new Error("Didn't throw");
      });
    });
  });
});
