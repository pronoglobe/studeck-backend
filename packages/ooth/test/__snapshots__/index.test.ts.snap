// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ooth no session connect method after login can connect another strategy 1`] = `
Object {
  "_id": "<obfuscated>",
  "test": Object {
    "foo": 1,
  },
  "test2": Object {
    "baz": 45,
  },
}
`;

exports[`ooth no session connect method after login can log in again 1`] = `
Object {
  "_id": "<obfuscated>",
  "test": Object {
    "foo": 2,
  },
}
`;

exports[`ooth no session connect method after login can log in with other strategy 1`] = `
Object {
  "_id": "<obfuscated>",
  "test": Object {
    "foo": 1,
  },
  "test2": Object {},
}
`;

exports[`ooth no session connect method can log in 1`] = `
Object {
  "user": Object {
    "_id": "<obfuscated>",
    "test": Object {
      "foo": 1,
    },
  },
}
`;

exports[`ooth no session connect method can register 1`] = `
Object {
  "_id": "<obfuscated>",
  "test": Object {
    "foo": 1,
  },
}
`;

exports[`ooth no session methods fails with requireLogged 1`] = `
Object {
  "message": "Not logged in",
  "status": "error",
}
`;

exports[`ooth no session methods handle async errors 1`] = `
Object {
  "message": "Error message.",
  "status": "error",
}
`;

exports[`ooth no session methods handle errors 1`] = `
Object {
  "message": "Error message.",
  "status": "error",
}
`;

exports[`ooth no session methods handle method 1`] = `
Object {
  "message": "hi",
}
`;

exports[`ooth no session methods translates error 1`] = `
Object {
  "message": "French not logged",
  "status": "error",
}
`;

exports[`ooth session methods can do secondary auth with cookies 1`] = `
Object {
  "_id": "<obfuscated>",
  "foo": Object {
    "x": 1,
  },
}
`;

exports[`ooth session methods can do secondary auth with cookies 2`] = `
Object {
  "_id": "<obfuscated>",
  "foo": Object {
    "x": 1,
  },
}
`;

exports[`ooth session methods can log out 1`] = `
Object {
  "_id": "<obfuscated>",
  "foo": Object {
    "x": 1,
  },
}
`;

exports[`ooth session methods can log out 2`] = `Object {}`;

exports[`ooth session methods can log out 3`] = `
Object {
  "message": "Not logged in",
  "status": "error",
}
`;

exports[`ooth session methods fails without cookies 1`] = `
Object {
  "message": "Not logged in",
  "status": "error",
}
`;
